// markdown-it/lib/rules_core/replacements.mjs
// Simple typographic replacements
//
// (c) (C) → ©
// (tm) (TM) → ™
// (r) (R) → ®
// +- → ±
// ... → … (also ?.... → ?.., !.... → !..)
// ???????? → ???, !!!!! → !!!, `,,` → `,`
// -- → &ndash;, --- → &mdash;
//
// TODO:
// - fractionals 1/2, 1/4, 3/4 -> ½, ¼, ¾
// - multiplications 2 x 4 -> 2 × 4

// Regex to detect "rare" replacements in text.
// Matches: "+-", ".." (two or more dots), "????" (four or more ?),
// "!!!!" (four or more !), and "--" (double dash).
// Note: we removed `,,` here to preserve double commas.
const RARE_RE = /\+-|\.\.|,,|\?\?\?\?|!!!!|--/

// Regex to detect scoped abbreviations like (c), (tm), (r).
// TEST_RE is case-insensitive, used just to check if content has them.
const SCOPED_ABBR_TEST_RE = /\((c|tm|r)\)/i

// Actual replacement regex, global + case-insensitive.
const SCOPED_ABBR_RE = /\((c|tm|r)\)/ig

// Mapping table for abbreviations.
const SCOPED_ABBR = {
  c: '©',
  r: '®',
  tm: '™'
}

// Replacement function: takes match and abbreviation name,
// returns the corresponding symbol.
function replaceFn(match, name) {
  return SCOPED_ABBR[name.toLowerCase()]
}

// Replace scoped abbreviations inside inline tokens.
function replace_scoped(inlineTokens) {
  let inside_autolink = 0

  for (let i = inlineTokens.length - 1; i >= 0; i--) {
    const token = inlineTokens[i]

    // Only process plain text, not inside auto-links.
    if (token.type === 'text' && !inside_autolink) {
      token.content = token.content.replace(SCOPED_ABBR_RE, replaceFn)
    }

    // Track whether we are inside an auto-link.
    if (token.type === 'link_open' && token.info === 'auto') {
      inside_autolink--
    }
    if (token.type === 'link_close' && token.info === 'auto') {
      inside_autolink++
    }
  }
}

// Replace "rare" typographic sequences.
function replace_rare(inlineTokens) {
  let inside_autolink = 0

  for (let i = inlineTokens.length - 1; i >= 0; i--) {
    const token = inlineTokens[i]

    if (token.type === 'text' && !inside_autolink) {
      if (RARE_RE.test(token.content)) {
        token.content = token.content
          // "+-" → ±
          .replace(/\+-/g, '±')
          // Two or more dots → ellipsis (…)
          // But if preceded by ? or !, collapse to ".."
          .replace(/\.{2,}/g, '…').replace(/([?!])…/g, '$1..')
          // Four or more ? or ! → collapse to three
          .replace(/([?!]){4,}/g, '$1$1$1')
          // Multiple commas → single comma
          // (comment out if you want to keep ,, untouched)
          //.replace(/,{2,}/g, ',')
          // --- → em-dash (—)
          .replace(/(^|[^-])---(?=[^-]|$)/mg, '$1\u2014')
          // -- surrounded by spaces → en-dash (–)
          .replace(/(^|\s)--(?=\s|$)/mg, '$1\u2013')
          // -- between non-space chars → en-dash (–)
          .replace(/(^|[^-\s])--(?=[^-\s]|$)/mg, '$1\u2013')
      }
    }

    // Track auto-link context.
    if (token.type === 'link_open' && token.info === 'auto') {
      inside_autolink--
    }
    if (token.type === 'link_close' && token.info === 'auto') {
      inside_autolink++
    }
  }
}

// Main entry point: runs replacements if typographer is enabled.
export default function replace(state) {
  let blkIdx

  if (!state.md.options.typographer) { return }

  for (blkIdx = state.tokens.length - 1; blkIdx >= 0; blkIdx--) {
    if (state.tokens[blkIdx].type !== 'inline') { continue }

    // Replace abbreviations like (c), (tm), (r).
    if (SCOPED_ABBR_TEST_RE.test(state.tokens[blkIdx].content)) {
      replace_scoped(state.tokens[blkIdx].children)
    }

    // Replace rare sequences like "+-", "..", "!!!!", "--".
    if (RARE_RE.test(state.tokens[blkIdx].content)) {
      replace_rare(state.tokens[blkIdx].children)
    }
  }
}
