// markdown-it-abcmusic/dist/index.js
import { JSDOM as f } from "jsdom";
import w from "abcjs";
class p {
  dom;
  constructor() {
    this.dom = new f(
      '<!DOCTYPE html><html><body><div id="paper"></div></body></html>',
      { pretendToBeVisual: !0 }
    ), global.window = this.dom.window, global.document = this.dom.window.document, global.HTMLElement = this.dom.window.HTMLElement, global.SVGElement = this.dom.window.SVGElement;
  }
  render(e) {
    const o = this.dom.window.document.getElementById("paper");
    o.innerHTML = "", w.renderAbc(o, e, {
      scale: 1
    });
    const t = o.querySelector("svg");
    if (!t)
      throw new Error("abcjs failed to produce SVG");
    return t.outerHTML;
  }
}
class k {
  renderer = new p();
  install(e) {
    e.block.ruler.before("fence", "abc_block", this.blockRule.bind(this), {
      alt: ["paragraph", "reference", "blockquote"]
    });
  }
  blockRule(e, r, o, t) {
    const d = e.bMarks[r] + e.tShift[r], u = e.eMarks[r];
    if (!e.src.slice(d, u).startsWith(":::abc"))
      return !1;
    let n = r + 1, l = [];
    for (; n < o; ) {
      const b = e.bMarks[n] + e.tShift[n], h = e.eMarks[n], s = e.src.slice(b, h);
      if (s.trim() === ":::")
        break;
      l.push(s), n++;
    }
    if (n >= o)
      return !1;
    if (t)
      return !0;
    const m = l.join(`
`), a = this.renderer.render(m), i = e.push("html_block", "", 0);
    return i.block = !0, i.content = `<div class="abc-music">${a}</div>
`, e.line = n + 1, !0;
  }
}
const g = "0.0.12";
function S() {
  console.log(`markdown-it-abcmusic v${g}`);
}
function T(c) {
  new k().install(c), console.log(S());
}
export {
  g as VERSION,
  T as default,
  S as printVersion
};
