// markdown-it/lib/rules_inline/smallcaps.mjs
// Process |this| and §that§
//

// Insert each marker as a separate text token, and add it to delimiter list
//
function smallcaps_tokenize(state, silent) {
  const start = state.pos
  const marker = state.src.charCodeAt(start)

  if (silent) { return false }

  if (marker !== 0x7C /* | */ && marker !== 0xA7 /* § */) { return false }

  const scanned = state.scanDelims(state.pos, false)

  for (let i = 0; i < scanned.length; i++) {
    const token = state.push('text', '', 0)
    token.content = String.fromCharCode(marker)

    state.delimiters.push({
      marker,
      length: scanned.length,
      token: state.tokens.length - 1,
      end: -1,
      open: scanned.can_open,
      close: scanned.can_close
    })
  }

  state.pos += scanned.length

  return true
}

function postProcess(state, delimiters) {
  const max = delimiters.length

  for (let i = max - 1; i >= 0; i--) {
    const startDelim = delimiters[i]

    if (startDelim.marker !== 0x7C /* | */ && startDelim.marker !== 0xA7 /* § */) {
      continue
    }

    // Process only opening markers
    if (startDelim.end === -1) {
      continue
    }

    const endDelim = delimiters[startDelim.end]

    // If the previous delimiter has the same marker and is adjacent to this one,
    // merge those into one strong delimiter.
    //
    // `<span><span>whatever</span></span>` -> `<span style="color:red">whatever</span>`
    //
    const isStrong = i > 0 &&
      delimiters[i - 1].end === startDelim.end + 1 &&
      delimiters[i - 1].marker === startDelim.marker &&
      delimiters[i - 1].token === startDelim.token - 1 &&
      delimiters[startDelim.end + 1].token === endDelim.token + 1

    const ch = String.fromCharCode(startDelim.marker)

    const token_o = state.tokens[startDelim.token]
    token_o.type = isStrong ? 'smallcaps_strong_open' : 'smallcaps_open'
    token_o.tag = 'span'
    token_o.nesting = 1
    token_o.markup = isStrong ? ch + ch : ch
    token_o.attrs = isStrong
      ? [['style', 'font-variant: petite-caps;']]
      : [['style', 'font-variant: small-caps;']]
    token_o.content = ''

    const token_c = state.tokens[endDelim.token]
    token_c.type = isStrong ? 'smallcaps_strong_close' : 'smallcaps_close'
    token_c.tag = 'span'
    token_c.nesting = -1
    token_c.markup = isStrong ? ch + ch : ch
    token_c.content = ''

    if (isStrong) {
      state.tokens[delimiters[i - 1].token].content = ''
      state.tokens[delimiters[startDelim.end + 1].token].content = ''
      i--
    }
  }
}

// Walk through delimiter list and replace text tokens with tags
//
function smallcaps_post_process(state) {
  const tokens_meta = state.tokens_meta
  const max = state.tokens_meta.length

  postProcess(state, state.delimiters)

  for (let curr = 0; curr < max; curr++) {
    if (tokens_meta[curr] && tokens_meta[curr].delimiters) {
      postProcess(state, tokens_meta[curr].delimiters)
    }
  }
}

export default {
  tokenize: smallcaps_tokenize,
  postProcess: smallcaps_post_process
}
